types/validator_set.go

```diff
@@ -78,44 +78,52 @@ func (vals *ValidatorSet) IncrementProposerPriority(times int) {
 		panic("Cannot call IncrementProposerPriority with non-positive times")
 	}
 
-	const shiftEveryNthIter = 10
+	// cap the difference between priorities to be proportional to 2*totalPower by
```

Comment:
As imbalance happens only upon validator set changes, I would suggest to move this code to `execution.go/updateValidators` and/or `validator_set.go/Add, validator_set.go/Remove`. Idea is to to averaging and normalisation only when val set is changed and then have "plain" proposer selection at the beginning of each round. As this might involve significant changes maybe it's better to address this in a separate issue/PR.

Reply:
Thanks for the clarification. That makes sense for the normalisation, not sure about the averaging/shifting but probably too.